sklearn/metrics/_classification.py

```diff
@@ -337,62 +337,76 @@ def confusion_matrix(
     (np.int64(0), np.int64(2), np.int64(1), np.int64(1))
     """
     y_true, y_pred = attach_unique(y_true, y_pred)
+    xp, _, device_ = get_namespace_and_device(y_true, y_pred, labels, sample_weight)
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     if y_type not in ("binary", "multiclass"):
         raise ValueError("%s is not supported" % y_type)
 
     if labels is None:
         labels = unique_labels(y_true, y_pred)
     else:
-        labels = np.asarray(labels)
+        labels = xp.asarray(labels)
         n_labels = labels.size
         if n_labels == 0:
             raise ValueError("'labels' should contains at least one label.")
         elif y_true.size == 0:
-            return np.zeros((n_labels, n_labels), dtype=int)
-        elif len(np.intersect1d(y_true, labels)) == 0:
+            return xp.zeros((n_labels, n_labels), dtype=xp.int64, device=device_)
+        elif not xp.isin(labels, y_true).any():
             raise ValueError("At least one label specified must be in y_true")
 
     if sample_weight is None:
-        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
+        sample_weight = xp.ones(y_true.shape[0], dtype=xp.int64, device=device_)
     else:
-        sample_weight = np.asarray(sample_weight)
+        sample_weight = xp.asarray(sample_weight, device=device_)
 
     check_consistent_length(y_true, y_pred, sample_weight)
 
-    n_labels = labels.size
+    # TODO: remove condition when torch supports the size attribute
+    if xp.__name__ == "array_api_compat.torch":
+        n_labels = xp.size(labels)
+    else:
+        n_labels = labels.size
     # If labels are not consecutive integers starting from zero, then
     # y_true and y_pred must be converted into index form
     need_index_conversion = not (
-        labels.dtype.kind in {"i", "u", "b"}
-        and np.all(labels == np.arange(n_labels))
-        and y_true.min() >= 0
-        and y_pred.min() >= 0
+        xp.isdtype(labels.dtype, ("signed integer", "unsigned integer", "bool"))
+        and xp.all(labels == xp.arange(n_labels, device=device_))
+        and xp.min(y_true) >= 0
+        and xp.min(y_pred) >= 0
     )
     if need_index_conversion:
         label_to_ind = {y: x for x, y in enumerate(labels)}
-        y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-        y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+        y_pred = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_pred], device=device_
+        )
+        y_true = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_true], device=device_
+        )
```

Comment:
This is also only tested for ndarrays, because of the way our tests are written. It should work for the other array libraries that we currently integrate, but I feel we should in fact test this part of the code?

Reply:
It fails, I will fix this and add a test.

Reply:
I have asked [here](https://github.com/data-apis/array-api-strict/issues/109) and they have explained me how to deal with it. It is fixed and I have added a test.

(I have also tried to re-factor, but not successfully yet and at least for this PR, it will stay as it is.)

Reply:
Not necessarily for this PR but maybe we could use [`unique_inverse`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.unique_inverse.html#array_api.unique_inverse)

something like this would maybe work:
```py
_, y_pred = xp.unique_inverse(y_pred)
```

Maybe a clearer example with array-api-strict:
```py

In [1]: import array_api_strict
   ...: arr = array_api_strict.asarray([1, 3, 4, 1, 3, 5])
   ...: unique_values, labels = array_api_strict.unique_inverse(arr)
   ...: labels
   ...: 
Out[1]: Array([0, 1, 2, 0, 1, 3], dtype=array_api_strict.int64)
```

Reply:
Actually, looking a bit further, it looks even like we have `sklearn.utils._encode._unique(return_inverse=True)` that seems to have array API support already ...

Reply:
Oh, that looks pretty good. I will try it tomorrow. :)

Reply:
Not for this PR, but for information, @lesteve:

I have tried to make it work with `unique_inverse` and a few other approaches, but I find it pretty complex to handle the lookup and substitution with only array operations, especially with strings allowed as dtypes. Maybe someone else wants to give it a try. I won't try any further, but I can share my experience, if interested.

In this PR, we're fine with the status quo of using the mapping.